include/CL/cl_exp_tensor.h

```diff
@@ -0,0 +1,230 @@
+ * Copyright (c) 2022-2024 Henry Linjamäki, Michal Babej / Intel Finland Oy
+ *
+ * PoCL-specific proof-of-concept (draft) of Defined Builtin Kernels extension.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a
+ * copy of this software and/or associated documentation files (the
+ * "Materials"), to deal in the Materials without restriction, including
+ * without limitation the rights to use, copy, modify, merge, publish,
+ * distribute, sublicense, and/or sell copies of the Materials, and to
+ * permit persons to whom the Materials are furnished to do so, subject to
+ * the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included
+ * in all copies or substantial portions of the Materials.
+ *
+ * MODIFICATIONS TO THIS FILE MAY MEAN IT NO LONGER ACCURATELY REFLECTS
+ * KHRONOS STANDARDS. THE UNMODIFIED, NORMATIVE VERSIONS OF KHRONOS
+ * SPECIFICATIONS AND HEADER INFORMATION ARE LOCATED AT
+ *    https://www.khronos.org/registry/
+ *
+ * THE MATERIALS ARE PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+ * IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+ * CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+ * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+ * MATERIALS OR THE USE OR OTHER DEALINGS IN THE MATERIALS.
+ ******************************************************************************/
+
+
+#ifndef OPENCL_EXP_TENSOR_H
+#define OPENCL_EXP_TENSOR_H
+#include <CL/cl.h>
+
+/* types for describing dimensions & stride */
+typedef cl_ulong cl_tensor_shape;
+typedef cl_ulong cl_tensor_stride;
+typedef cl_uint cl_tensor_dim;
+
+typedef cl_uint cl_tensor_datatype;
+/* cl_tensor_datatype values. List of the datatypes for Tensor data */
+#define CL_TENSOR_DTYPE_UNKNOWN 0
+#define CL_TENSOR_DTYPE_FP64 1
+#define CL_TENSOR_DTYPE_INT64 2
+#define CL_TENSOR_DTYPE_UINT64 3
+#define CL_TENSOR_DTYPE_FP32 4
+#define CL_TENSOR_DTYPE_INT32 5
+#define CL_TENSOR_DTYPE_UINT32 6
+#define CL_TENSOR_DTYPE_FP16 7
+#define CL_TENSOR_DTYPE_INT16 8
+#define CL_TENSOR_DTYPE_UINT16 9
+#define CL_TENSOR_DTYPE_FP8 10
+#define CL_TENSOR_DTYPE_INT8 11
+#define CL_TENSOR_DTYPE_UINT8 12
+#define CL_TENSOR_DTYPE_INT4 13
+#define CL_TENSOR_DTYPE_UINT4 14
+#define CL_TENSOR_DTYPE_LAST 15
+
+/* cl_tensor_datatype_union is used to pass POD data to DBKs (like Alpha & Beta
+ * parameters to the GEMM); the actual type used is implied. In case of
+ * GEMM, the datatype is implied the same as COut Tensor datatype.
+ * TODO: this could be done with passing a void* pointer, however
+ * this avoids the need for malloc & memcpy */
+typedef union
+{
+  cl_char c;
+  cl_short s;
+  cl_int i;
+  cl_long l;
+  cl_half h;
+  cl_float f;
+  cl_double d;
+  char raw[sizeof (cl_double)];
+} cl_tensor_datatype_union;
```

Comment:
What if we need to pass objects larger than `sizeof(cl_double)` in the future? We can’t update this union due to potential ABI breakage.